### PR TITLE
Rebuild changed components only

### PR DESCRIPTION
**Pull request by nilsgstrabo**
### Rebuild changed components
A component is build if:
- One or more environments does not have an active RadixDeployment.
- RadixApplication.Spec sha256 does not match the value stored in active RadixDeployment annotation for environment to build
- BuildSecrets is defined in RadixApplication and sha256 of Secret.Data does not match the value stored in active RadixDeployment annotation for environment to build
- The component is listed in `PrepareBuildContext.EnvironmentsToBuild` returned from the prepare-pipeline step.

If a component does not need to be rebuilt, the image is taken from previous RadixDeployment for the environment.

### Misc other changes
- Moved utils package to internal
- Reactored and moved unit tests

### CODE REVIEW

File: pipeline-runner/model/pipelineInfo.go [hunk 1]
```diff
@@ -217,145 +198,6 @@ func (info *PipelineInfo) SetGitAttributes(gitCommitHash, gitTags string) {
 }
 
 // IsDeployOnlyPipeline Determines if the pipeline is deploy-only
```
Comment (satr) [inline pipeline-runner/model/pipelineInfo.go:200]: Inconsistent method name in the comment

File: pipeline-runner/steps/apply_radixconfig.go [hunk 2]
```diff
@@ -74,21 +87,31 @@ func (cli *ApplyConfigStepImplementation) Run(pipelineInfo *model.PipelineInfo)
...
+	if err := cli.setBuildSecret(pipelineInfo); err != nil {
+		return err
+	}
+
+	if err := cli.setBuildAndDeployImages(pipelineInfo); err != nil {
+		return err
+	}
+
+	if pipelineInfo.IsPipelineType(radixv1.Deploy) {
+		if len(pipelineInfo.BuildComponentImages) > 0 {
```
Comment (satr) [inline pipeline-runner/steps/apply_radixconfig.go:109]: One "if"?
```
if pipelineInfo.IsPipelineType(radixv1.Deploy) && 
len(pipelineInfo.BuildComponentImages) > 0 {
```

File: pkg/apis/deployment/deployment.go [hunk 3]
```diff
@@ -478,13 +478,25 @@ func (deploy *Deployment) garbageCollectAuxiliaryResources() error {
...
 	}
+	annotations := map[string]string{
+		kube.RadixBranchAnnotation:  branch,
+		kube.RadixGitTagsAnnotation: gitTags,
+		kube.RadixCommitAnnotation:  commitID,
+	}
+	if len(buildSecretHash) > 0 {
+		annotations[kube.RadixBuildSecretHash] = buildSecretHash
+	}
+
```
Comment (Richard87) [inline pkg/apis/deployment/deployment.go:496]: Is there any pros/cons to use `len(string) > 0` instead of `string != ""`?

**Review by Richard87 — DISMISSED** [review]
Ett par spørsmål :)

File: pipeline-runner/model/pipelineInfo.go (hunk 1 above)
Comment (nilsgstrabo) [reply pipeline-runner/model/pipelineInfo.go]: fixed

File: pipeline-runner/steps/apply_radixconfig.go (hunk 2 above)
Comment (nilsgstrabo) [reply pipeline-runner/steps/apply_radixconfig.go]: fixed

File: pipeline-runner/model/pipelineInfo.go [hunk 4]
```diff
@@ -2,53 +2,43 @@ package model
...
+	RadixApplication  *radixv1.RadixApplication
+	BuildSecret       *corev1.Secret
 	PipelineArguments PipelineArguments
 	Steps             []Step
 
 	// Temporary data
 	RadixConfigMapName string
 	GitConfigMapName   string
-	TargetEnvironments map[string]bool
+	TargetEnvironments map[string]bool // TODO: Change this to []string
```
[the remark this answers is not in the document]
Comment (nilsgstrabo) [reply pipeline-runner/model/pipelineInfo.go]: fixed

File: pkg/apis/deployment/deployment.go (hunk 3 above)
Comment (nilsgstrabo) [reply pkg/apis/deployment/deployment.go:496]: Nah, don't think so. We usually use len() to check length

**Review by Richard87 — APPROVED** [review]
godkjent

**Review by satr — APPROVED** [review]

**Pull request merged**